8312522: Implementation of Foreign Function & Memory API

JornVernee opened the pull request:
This patch contains the implementation of the foreign linker & memory API JEP for Java 22. The initial patch is composed of commits brought over directly from the [panama-foreign repo](https://github.com/openjdk/panama-foreign). The main changes found in this patch come from the following PRs:

1. https://github.com/openjdk/panama-foreign/pull/836 Where previous iterations supported converting Java strings to and from native strings in the UTF-8 encoding, we've extended the supported encoding to all the encodings found in the `java.nio.charset.StandardCharsets` class.
2. https://github.com/openjdk/panama-foreign/pull/838 We dropped the `MemoryLayout::sequenceLayout` factory method which inferred the size of the sequence to be `Long.MAX_VALUE`, as this led to confusion among clients. A client is now required to explicitly specify the sequence size.
3. https://github.com/openjdk/panama-foreign/pull/839 A new API was added: `Linker::canonicalLayouts`, which exposes a map containing the platform-specific mappings of common C type names to memory layouts.
4. https://github.com/openjdk/panama-foreign/pull/840 Memory access varhandles, as well as byte offset and slice handles derived from memory layouts, now feature an additional 'base offset' coordinate that is added to the offset computed by the handle. This allows composing these handles with other offset computation strategies that may not be based on the same memory layout. This addresses use-cases where clients are working with 'dynamic' layouts, whose size might not be known statically, such as variable length arrays, or variable size matrices.
5. https://github.com/openjdk/panama-foreign/pull/841 Remove this now redundant API. Clients can simply use the difference between the base address of two memory segments.
6. https://github.com/openjdk/panama-foreign/pull/845 Disambiguate uses of `SegmentAllocator::allocateArray`, by renaming methods that both allocate + initialize memory segments to `allocateFrom`. (see the original PR for the problematic case)
7. https://github.com/openjdk/panama-foreign/pull/846 Improve the documentation for variadic functions.
8. https://github.com/openjdk/panama-foreign/pull/849 Several test fixes to make sure the `jdk_foreign` tests can pass on 32-bit machines, taking linux-x86 as a test bed.
9. https://github.com/openjdk/panama-foreign/pull/850 Make the linker API required. The `Linker::nativeLinker` method is not longer allowed to throw an `UnsupportedOperationException` on unsupported platforms. All tests are turned on by default, instead of being skipped when the linker is not present.
10. https://github.com/openjdk/panama-foreign/pull/851 Minor code tweaks to ensure the JIT can constant fold through native access checks if the accessing class is statically known (see commit/original PR for changes).
11. https://github.com/openjdk/panama-foreign/pull/853 Remove all the `@PreviewFeature` annotations from the API. Update all `@since` tags in the Javadoc to `@since 22` (per [JEP 12](https://bugs.openjdk.org/browse/JDK-8195734)). Update tests and benchmarks to no longer build and run using `--enable-preview` (or the `@enablePreview` jtreg tag).

I want to call out in particular that this patch finalizes the FFM API (by moving it out of preview), and requires all JDK implementations to implement it. Most ports already have full FFM API support. The ones that are missing are: s390 ([currently under review](https://github.com/openjdk/jdk/pull/14801)), windows-x86 (deprecated for removal), and linux-86 & arm32 (which can both be implemented using the fallback linker [1](https://github.com/openjdk/panama-foreign/pull/770) [2](https://mail.openjdk.org/pipermail/porters-dev/2023-March/000753.html)).

Testing: tier 1-5

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a JEP request to be targeted
- [x] Change requires CSR request [JDK-8312523](https://bugs.openjdk.org/browse/JDK-8312523) to be approved

### Issues
 * [JDK-8312522](https://bugs.openjdk.org/browse/JDK-8312522): Implementation of Foreign Function &amp; Memory API (**Enhancement** - P3)
 * [JDK-8310626](https://bugs.openjdk.org/browse/JDK-8310626): JEP 454: Foreign Function &amp; Memory API (**JEP**)
 * [JDK-8312523](https://bugs.openjdk.org/browse/JDK-8312523): Implementation of Foreign Function &amp; Memory API (**CSR**)


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [b0a1abaf](https://git.openjdk.org/jdk/pull/15103/files/b0a1abafd2386a7609f608fd19811ca685e1089a)
 * [Paul Sandoz](https://openjdk.org/census#psandoz) (@PaulSandoz - **Reviewer**) ⚠️ Review applies to [0244845a](https://git.openjdk.org/jdk/pull/15103/files/0244845abbf06adecd9a2e4217bc26f41825f875)
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to [1f3248df](https://git.openjdk.org/jdk/pull/15103/files/1f3248df59823de15c15d58d8da9093f32405281)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [2bc0a650](https://git.openjdk.org/jdk/pull/15103/files/2bc0a65086d0cfc214b59a18b821929b72a83a7f)


### Contributors
 * Maurizio Cimadamore `<mcimadamore@openjdk.org>`
 * Jorn Vernee `<jvernee@openjdk.org>`
 * Per Minborg `<pminborg@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15103/head:pull/15103` \
`$ git checkout pull/15103`

Update a local copy of the PR: \
`$ git checkout pull/15103` \
`$ git pull https://git.openjdk.org/jdk.git pull/15103/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15103`

View PR using the GUI difftool: \
`$ git pr show -t 15103`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15103.diff">https://git.openjdk.org/jdk/pull/15103.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15103#issuecomment-1660591225)

JornVernee commented:
/label remove shenandoah

JornVernee commented:
/label add hotspot-compiler

JornVernee commented:
/contributor add @mcimadamore 
/contributor add @JornVernee

JornVernee commented:
/jep JDK-8310626

JornVernee commented:
Open build question: the `jdk_foreign` tests will fail if a JDK does not have a full linker port, and is not built with `--enable-fallback-linker`. If the API is to become required, as proposed by this PR, then should the build also require the fallback linker by default on platforms that do not have a full linker port?

ExE-Boss commented on src/java.base/share/classes/jdk/internal/foreign/abi/fallback/FallbackLinker.java:
```diff
@@ -283,4 +293,41 @@ private static Object readValue(MemorySegment seg, MemoryLayout layout) {
+
+    @Override
+    public Map<String, MemoryLayout> canonicalLayouts() {
+        return CANONICAL_LAYOUTS;
```
`CANONICAL_LAYOUTS` is set to a `HashMap`, which is not unmodifiable.
```suggestion
        return Collections.unmodifiableMap(CANONICAL_LAYOUTS);
```

JornVernee replied:
Good catch. I think the right fix is to update FallbackLinker though. The other ports already use `Map.of`.

JornVernee commented:
FWIW, GHA linux-x86 currently fail tests, since it is not being build with `--enable-fallback-linker`. I think this could be easily resolved, but attempting this currently fails when trying to install GCC:

```
Package gcc-10-multilib is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source
```

This error occurs even on a plain master branch, so it doesn't seem to be caused by this PR.

erikj79 reviewed (commented):
Build change looks ok.

dholmes-ora reviewed (approved):
Hotspot test changes look fine. Thanks

dholmes-ora commented:
/reviewers needed 2

JornVernee commented:
Thanks for the reviews so far.

That also reminds me to mention: while there are a lot of files touched, most of the updates a simple 1 line changes to test files removing the `@enablePreview` tag from the jtreg test information. So, probably most of this is easy to review.

JornVernee commented:
Addressing some offline review comments. Two more changes have been added:
- Disallow combining the `captureCallState` and `isTrivial` (see https://github.com/openjdk/panama-foreign/pull/856)
- Rename `isTrivial` to `critical` (see https://github.com/openjdk/panama-foreign/pull/859)

bplb commented:
The few, simple NIO changes are fine.

liach commented:
Just curious, what's the rationale for finalizing the API when there are significant changes from the last preview?

uschindler commented:
> > Just curious, what's the rationale for finalizing the API when there are significant changes from the last preview?
> 
> A preview API is finalized when it is ready. The preview process, as outlined by [JEP 12](https://bugs.openjdk.org/browse/JDK-8195734), does not place a mandate on the amount of changes that a JEP that finalizes a preview API should or should not contain. It only requires that the changes since the last preview iteration are noted (which we have done). Though, the amount of changes can be used to inform the decision to finalize. We feel that the FFM API is ready for finalization, and does not require another round of preview.
> 
> In this case in particular: previous iterations contained significant changes to the API, including re-shuffling of some of the core APIs. (See e.g. [#13079 (comment)](https://github.com/openjdk/jdk/pull/13079#issuecomment-1476648707)) In contrast this JEP contains mostly superficial changes to the API, that are not likely to impact how a client would write a program using the FFM API.

In addition if somebody wrote code against the preview API, heshe needs to update it anyways because the class files are marked by preview flag. So all is fine. We just make API ready to use for everybody and therefor all early adopters need to adapt. t won't affect anybox else.

To me it would be strange if code goes out of preview without changes, because if there are no changes why was it in preview then in the last JDK feature version?

JornVernee commented:
/contributor add @minborg

PaulSandoz reviewed (commented):
Recommend you do a sweep through the code for unused imports and fields and any rogue `@since` in the internal classes.

JornVernee commented on src/java.base/share/classes/java/lang/foreign/MemoryLayout.java:
```diff
@@ -351,71 +428,93 @@ default MethodHandle byteOffsetHandle(PathElement... elements) {
-     * and {@code s_1}, {@code s_2}, ... {@code s_n} are <em>static</em> stride constants which are derived from
-     * the layout path.
+     * {@snippet lang = "java":
+     * offset = this.offsetHandle(P).invokeExact(B, I1, I2, ... In);
```
Good catch. The `O` later one was added separately, hence they are out of sync.

JornVernee commented on src/java.base/share/classes/java/lang/invoke/MethodHandles.java:
```diff
@@ -7951,85 +7951,6 @@ private static MethodType tableSwitchChecks(MethodHandle defaultCase, MethodHand
-     * @see MemoryLayout#varHandle(MemoryLayout.PathElement...)
-     * @since 19
-     */
-    @PreviewFeature(feature=PreviewFeature.Feature.FOREIGN)
```
Thanks. I did a sweep of the foreign packages and there don't seem to be any other uses. There were also some other unused imports in this file that I'll remove.

TheRealMDoerr commented:
Please adapt the new `LinuxPPC64Linker` when merging:
```diff
diff --git a/src/java.base/share/classes/jdk/internal/foreign/abi/ppc64/linux/LinuxPPC64Linker.java b/src/java.base/share/classes/jdk/inter
nal/foreign/abi/ppc64/linux/LinuxPPC64Linker.java
index 150687d4078..7cf2d524bff 100644
--- a/src/java.base/share/classes/jdk/internal/foreign/abi/ppc64/linux/LinuxPPC64Linker.java
+++ b/src/java.base/share/classes/jdk/internal/foreign/abi/ppc64/linux/LinuxPPC64Linker.java
@@ -27,15 +27,22 @@
 
 import jdk.internal.foreign.abi.AbstractLinker;
 import jdk.internal.foreign.abi.LinkerOptions;
+import jdk.internal.foreign.abi.SharedUtils;
 import jdk.internal.foreign.abi.ppc64.CallArranger;
 
 import java.lang.foreign.FunctionDescriptor;
+import java.lang.foreign.MemoryLayout;
+import java.lang.foreign.ValueLayout;
 import java.lang.invoke.MethodHandle;
 import java.lang.invoke.MethodType;
 import java.nio.ByteOrder;
+import java.util.Map;
 
 public final class LinuxPPC64Linker extends AbstractLinker {
 
+    static final Map<String, MemoryLayout> CANONICAL_LAYOUTS =
+        SharedUtils.canonicalLayouts(ValueLayout.JAVA_LONG, ValueLayout.JAVA_LONG, ValueLayout.JAVA_INT);
+
     public static LinuxPPC64Linker getInstance() {
         final class Holder {
             private static final LinuxPPC64Linker INSTANCE = new LinuxPPC64Linker();
@@ -62,4 +69,9 @@ protected UpcallStubFactory arrangeUpcall(MethodType targetType, FunctionDescrip
     protected ByteOrder linkerByteOrder() {
         return ByteOrder.BIG_ENDIAN;
     }
+
+    @Override
+    public Map<String, MemoryLayout> canonicalLayouts() {
+        return CANONICAL_LAYOUTS;
+    }
 }
```
The `test/jdk/java/foreign` tests have passed on both, linux PPC64 and PPC64le.

JornVernee commented on src/java.base/share/classes/java/lang/foreign/Linker.java:
```diff
@@ -122,7 +123,33 @@
+ *     <li>{@code wchar_t}</li>
+ *     <li>{@code void*}</li>
+ * </ul>
+ * As noted above, the specific canonical layout associated with each type can vary, depending on the data model
+ * supported by a given ABI. For instance, the C type {@code long} maps to the layout constant {@link ValueLayout#JAVA_LONG}
+ * on Linux/x64, but maps to the layout constant {@link ValueLayout#JAVA_INT} on Windows/x64. Similarly, the C type
+ * {@code size_t} maps to the layout constant {@link ValueLayout#JAVA_LONG} on 64-bit platforms, but maps to the layout
+ * constant {@link ValueLayout#JAVA_INT} on 32-bit platforms.
+ * <p>
+ * A native linker typically does not provide canonical layouts for C's unsigned integral types. Instead, they are
+ * modelled using the canonical layouts associated with their corresponding signed integral types. For instance,
+ * the C type {@code unsigned long} maps to the layout constant {@link ValueLayout#JAVA_LONG} on Linux/x64, but maps to
+ * the layout constant {@link ValueLayout#JAVA_INT} on Windows/x64.
+ * <p>
+ * The following table shows some examples of how C types are modelled in Linux/x64 (all the examples provided
+ * here will assume these platform-dependent mappings):
```
There is this: https://refspecs.linuxbase.org/elf/x86_64-abi-0.99.pdf I'm not sure how stable this is. I don't think that website is an authoritative source. (at least, not to the degree it is for e.g. AArch64: https://github.com/ARM-software/abi-aa/blob/main/aapcs64/aapcs64.rst#595homogeneous-aggregates). Note also that is a draft. I think the final version is paywalled.

Alternatively, we could refer to the name only: "System V Application Binary Interface - AMD64 Architecture Processor Supplement" (or "x86-64 psABI")

Then people can google for themselves and find it. For instance, [this SO question](https://stackoverflow.com/a/40348010) points to a gitlab repo with a more up to date version: https://gitlab.com/x86-psABIs/x86-64-ABI/-/jobs/artifacts/master/raw/x86-64-ABI/abi.pdf?job=build

PaulSandoz replied:
Yeah, its hard to find the official and latest version. Referring to the full title will help.

JornVernee replied:
I've added the name now.

JornVernee commented:
@TheRealMDoerr Thanks for the patch, will include it.

JornVernee commented:
After discussing with Maurizio, I've added one more non-trivial change to this patch. This adds a new `SegmentAllocator::allocateFrom` overload which accepts a MemorySegment as an initializer. See the original PR: https://github.com/openjdk/panama-foreign/pull/878

The commit I've added also includes the changes from https://github.com/openjdk/panama-foreign/pull/855 which were required by the first patch. I had to massage the code a bit since the javadoc in the mainline slightly deviates from the one in the panama-foreign repo.

Please take a look at the changes here: https://github.com/openjdk/jdk/pull/15103/commits/55296527a029b80dd78a7d1aecb429e793d7d32e

ExE-Boss commented on src/java.base/share/classes/jdk/internal/foreign/NativeMemorySegmentImpl.java:
```diff
@@ -147,6 +147,14 @@ public void cleanup() {
+        try {
+            return UNSAFE.allocateMemory(size);
+        } catch (IllegalArgumentException ex) {
+            throw new OutOfMemoryError();
```
`OutOfMemoryError` should be updated to have the `Throwable`‑accepting constructor overloads, so that this can include the cause:
```suggestion
            throw new OutOfMemoryError(ex);
```

See also: https://github.com/openjdk/panama-foreign/pull/855#discussion_r1285058300

JornVernee replied:
I think enhancing `OutOfMemoryError` is a discussion to be had separately. Not as a part of this JEP.

ExE-Boss commented on src/java.base/share/classes/java/lang/foreign/Linker.java:
```diff
@@ -569,8 +569,12 @@ static Linker nativeLinker() {
-     * representing the target address of the foreign function is the {@link MemorySegment#NULL} address.
-     * The returned method handle will additionally throw {@link NullPointerException} if any argument passed to it is {@code null}.
+     * representing the target address of the foreign function is the {@link MemorySegment#NULL} address. If an argument
+     * is a {@link MemorySegment},whose corresponding layout is an {@linkplain GroupLayout group layout}, the linker might attempt to access the contents of the segment. As such, one of the exceptions specified by the
```
**Nit:**
```suggestion
     * is a {@link MemorySegment},whose corresponding layout is a {@linkplain GroupLayout group layout}, the linker might attempt to access the contents of the segment. As such, one of the exceptions specified by the
```

ExE-Boss commented on src/java.base/share/classes/jdk/internal/foreign/abi/fallback/FallbackLinker.java:
```diff
@@ -283,4 +293,44 @@ private static Object readValue(MemorySegment seg, MemoryLayout layout) {
+        return CANONICAL_LAYOUTS;
+    }
+
+    static final Map<String, MemoryLayout> CANONICAL_LAYOUTS;
+
+    static {
+        int wchar_size = LibFallback.wcharSize();
+        MemoryLayout wchartLayout = switch(wchar_size) {
+            case 2 -> JAVA_CHAR; // prefer JAVA_CHAR
+            default -> FFIType.layoutFor(wchar_size);
+        };
+
+        CANONICAL_LAYOUTS = Map.ofEntries(
```
`LibFallback::wcharSize()` and other getters for `LibFallback.NativeConstants` fields can throw an error when `LibFallback.SUPPORTED` is `false` due to the `fallbackLinker` library not being present, so this static initializer should be made into a method instead:
```suggestion
    static final Map<String, MemoryLayout> CANONICAL_LAYOUTS = initCanonicalLayouts();

    private static Map<String, MemoryLayout> initCanonicalLayouts() {
        if (!isSupported()) {
            return null;
        }

        int wchar_size = LibFallback.wcharSize();
        MemoryLayout wchartLayout = switch(wchar_size) {
            case 2 -> JAVA_CHAR; // prefer JAVA_CHAR
            default -> FFIType.layoutFor(wchar_size);
        };

        return Map.ofEntries(
```

JornVernee replied:
Good catch! I've chosen a slightly different solution though.

mcimadamore commented on src/java.base/share/classes/java/lang/foreign/Linker.java:
```diff
@@ -122,7 +123,34 @@
+ * the layout constant {@link ValueLayout#JAVA_INT} on Windows/x64.
+ * <p>
+ * The following table shows some examples of how C types are modelled in Linux/x64 according to the
+ * "System V Application Binary Interface - AMD64 Architecture Processor Supplement" (all the examples provided
```
I have seen some discussion on this and I agree an authoritative link does not exist (I have searched for it in the past). I guess I'm not super wild about also including "AMD64 Architecture Processor Supplement". E.g. IMHO "System V Application Binary Interface" is more than enough?

JornVernee replied:
Ok, I think I agree, given that we already name x64 as a platform.

mcimadamore commented on src/java.base/share/classes/java/lang/foreign/SegmentAllocator.java:
```diff
@@ -269,163 +298,192 @@ default MemorySegment allocate(ValueLayout.OfDouble layout, double value) {
+    }
+
+    /**
+     * {@return a new memory segment with a {@linkplain MemorySegment#byteSize() byteSize()} of
```
```suggestion
     * {@return a new memory segment with a {@linkplain MemorySegment#byteSize() byte size} of
```

mcimadamore replied:
Also, in the panama repo I see this:
```
Allocates a memory segment with the given layout and initializes it with the bytes in the provided source memory segment.
```
Which seems more correct - e.g. more consistent with other allocation methods, and also more succinct (note that the first sentence is really what shows up in the method summary javadoc, so there is a certain interest in providing a quick description of what the method does,

mcimadamore replied:
Panama repo change:
https://github.com/openjdk/panama-foreign/commit/06e2017883c939188103c4dd53185417a00b2921

But, this code was altered in a follow up merge - maybe the merge was problematic?

JornVernee replied:
This was changed in the main line repo as a result of: https://github.com/openjdk/jdk/pull/14997 Since all the other methods were using `{@return ...}` I changed this new overload to that style as well.

JornVernee replied:
I don't mind changing it back to the old style, but I think the style should be consistent for all the allocateFrom overloads? So, I'd have to change all of them back.

mcimadamore replied:
I forgot about the change that went into mainline. Do you have a link of the latest javadoc? I'd like to check how the method summary looks.

JornVernee replied:
Here you go: https://cr.openjdk.org/~jvernee/FFM_22_PR_v1/java.base/java/lang/foreign/SegmentAllocator.html#allocateFrom(java.lang.foreign.ValueLayout,java.lang.foreign.MemorySegment,java.lang.foreign.ValueLayout,long,long)

mcimadamore replied:
Ok, now I'm more convinced that the method summary really does look bad (or worse, compared to 20).

For instance [allocateFrom](https://cr.openjdk.org/~jvernee/FFM_22_PR_v1/java.base/java/lang/foreign/SegmentAllocator.html#allocateFrom(java.lang.foreign.ValueLayout.OfByte,byte...):

```
Returns a new memory segment with a byteSize() initialized with the provided E byte elements as specified by the provided layout (i.e. byte ordering, alignment and size).
```

(same is true for all the other array-accepting `allocateFrom` methods). This should be simplified to:

```
Returns a new memory segment initialized with the elements in the provided byte array.
```
(then, if we want to say that the initialization honors the endianness of the provided layout, we can do so in a followup para, but the method summary should be simple).

So, once all the array-accepting methods are fixed, the segment-accepting `allocateFrom` needs to be simplified to:

```
Returns a new memory segment initialized with the contents of the provided segment.
```

JornVernee replied:
I've updated the 'header' of these methods, and instead added a more elaborated comment in a second paragraph. See: https://github.com/openjdk/jdk/pull/15103/commits/49bdd953eaed8732b59a803516355abc8df31fa3

Here is the re-generated javadoc: https://cr.openjdk.org/~jvernee/FFM_22_PR/v2/java.base/java/lang/foreign/SegmentAllocator.html#allocateFrom(java.lang.foreign.ValueLayout.OfInt,int...)

mcimadamore commented on test/micro/org/openjdk/bench/java/lang/foreign/TestLoadBytes.java:
```diff
@@ -50,8 +50,7 @@
 @OutputTimeUnit(TimeUnit.NANOSECONDS)
 @Fork(value = 1, jvmArgsAppend = {
         "-Dforeign.restricted=permit",
```
This seems obsolete? Maybe check other files too

mcimadamore commented on test/micro/org/openjdk/bench/java/lang/foreign/LoopOverSlice.java:
```diff
@@ -49,7 +49,7 @@
 @State(org.openjdk.jmh.annotations.Scope.Thread)
 @OutputTimeUnit(TimeUnit.MILLISECONDS)
-@Fork(value = 3, jvmArgsAppend = { "--enable-preview", "--enable-native-access=ALL-UNNAMED" })
+@Fork(value = 3, jvmArgsAppend = { "--enable-native-access=ALL-UNNAMED" })
```
Is --enable-native-access needed?

mcimadamore commented on test/micro/org/openjdk/bench/java/lang/foreign/LoopOverNonConstant.java:
```diff
@@ -48,7 +48,7 @@
 @State(org.openjdk.jmh.annotations.Scope.Thread)
 @OutputTimeUnit(TimeUnit.MILLISECONDS)
-@Fork(value = 3, jvmArgsAppend = { "--enable-preview", "--enable-native-access=ALL-UNNAMED" })
+@Fork(value = 3, jvmArgsAppend = { "--enable-native-access=ALL-UNNAMED" })
```
Is native access needed?

JornVernee commented on test/micro/org/openjdk/bench/java/lang/foreign/AllocFromSliceTest.java:
```diff
@@ -0,0 +1,80 @@
+@Measurement(iterations = 10, time = 500, timeUnit = TimeUnit.MILLISECONDS)
+@State(org.openjdk.jmh.annotations.Scope.Thread)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@Fork(value = 3, jvmArgsAppend = { "--enable-native-access=ALL-UNNAMED" })
```
This class extends `CLayouts` which also creates some native method handles, so it's needed here.

liach commented:
Just curious, for `Enable-Native-Access`, if it's present on an automatic module `Automatic-Module-Name` jar, can it apply to only that automatic module instead of all unnamed modules?

JornVernee commented:
> Just curious, for `Enable-Native-Access`, if it's present on an automatic module `Automatic-Module-Name` jar, can it apply to only that automatic module instead of all unnamed modules?

No. It's only there for executable jars (run using `java -jar`), which are always placed on the class path.

JornVernee commented on src/java.base/share/classes/java/lang/foreign/Linker.java:
```diff
@@ -390,15 +405,11 @@
- * non-variadic functions by replacing the {@code ...} or empty formal parameter list with a list of <em>variadic parameters</em>
- * of a fixed number and type.
+ * Variadic functions are C functions which can accept a variable number and type of arguments. They are declared with a
+ * trailing ellipsis ({@code ...}) at the end of the formal parameter list, such as: {@code void foo(int x, ...);}.
```
Will drop the period

mcimadamore commented on src/java.base/share/classes/java/lang/foreign/SegmentAllocator.java:
```diff
@@ -268,164 +320,210 @@ default MemorySegment allocate(ValueLayout.OfDouble layout, double value) {
-     *  return seg;
+     * {@return a new memory segment initialized with the contents of the provided segment.}
+     * <p>
+     * The size of the allocated memory segment is the {@code elementLayout.byteSize() * elementCount}.
```
```suggestion
     * The size of the allocated memory segment is {@code elementLayout.byteSize() * elementCount}.
```

mcimadamore commented on src/java.base/share/classes/java/lang/foreign/SegmentAllocator.java:
```diff
@@ -268,164 +320,210 @@ default MemorySegment allocate(ValueLayout.OfDouble layout, double value) {
-     *
+     * {@return a new memory segment initialized with the elements in the provided byte array.}
+     * <p>
+     * The size of the allocated memory segment is the {@code elementLayout.byteSize() * elements.length}.
```
```suggestion
     * The size of the allocated memory segment is {@code elementLayout.byteSize() * elements.length}.
```

mcimadamore replied:
Here and also in all the other array-accepting methods.

PaulSandoz commented on src/java.base/share/classes/java/lang/Module.java:
```diff
@@ -320,13 +319,16 @@ void ensureNativeAccess(Class<?> owner, String methodName) {
-                        WARNING: %s has been called by %s
-                        WARNING: Use --enable-native-access=%s to avoid a warning for this module
-                        %n""", cls, mtd, mod, modflag);
+                        WARNING: %s has been called%s in %s
```
```suggestion
                        WARNING: %s has been called by %s in %s
```
?

JornVernee replied:
The current code does the right thing, since in some cases the caller is `null` and the second `%s` should expand to an empty string. So  in the `caller == null` case, the message becomes:

    Class::method has been called in an unnamed module

There was also an offline suggestion to change it to:

    Class::method has been called by code in an unnamed module
    
Which I think is a good idea.

PaulSandoz commented on src/java.base/share/classes/java/lang/foreign/Linker.java:
```diff
@@ -704,6 +731,8 @@ static Option firstVariadicArg(int index) {
          * }
          * }
          *
+         * @apiNote This linker option can not be combined with {@link #critical}.
+         *
```
That seems more specification (that can be asserted on) then an informative note.

JornVernee replied:
True. I fold it into the main spec body.

PaulSandoz commented on src/java.base/share/classes/jdk/internal/foreign/NativeMemorySegmentImpl.java:
```diff
@@ -147,6 +147,14 @@ public void cleanup() {
 
+    private static long allocateMemoryWrapper(long size) {
+        try {
+            return UNSAFE.allocateMemory(size);
```
Since we now zero memory only when needed we should test very carefully.

JornVernee replied:
Yes. The `makeNativeSegment` is currently only called from ArenaImpl, which is also responsible for zeroing the memory.

I'll rename `makeNativeSegment` to `makeNativeSegmentNoZeroing` to make it extra clear for callers that memory will not be zeroed.

PaulSandoz commented on src/java.base/share/classes/java/lang/foreign/Linker.java:
```diff
@@ -29,14 +29,12 @@
-import jdk.internal.reflect.Reflection;
 
 import java.lang.invoke.MethodHandle;
 import java.nio.ByteOrder;
```
Unused?

JornVernee replied:
Yes, I'll remove it.

PaulSandoz reviewed (approved):
This looks good.

In the implementation the functional interfaces `BindingInterpreter.StoreFunc/LoadFunc` are package private, but are used in internal public signatures. This was previously like this and it's not a big deal but i recommend making those interfaces public.

We can also remove `@enablePreview` from `IndirectVarHandleTest`

JornVernee commented:
> In the implementation the functional interfaces `BindingInterpreter.StoreFunc/LoadFunc` are package private, but are used in internal public signatures. This was previously like this and it's not a big deal but i recommend making those interfaces public.

This was fixed in the panama-foreign repo: https://github.com/openjdk/panama-foreign/pull/891

I'll add that commit to this patch as well.

> We can also remove `@enablePreview` from `IndirectVarHandleTest`

Good catch!

AlanBateman commented on src/java.base/share/classes/sun/launcher/LauncherHelper.java:
```diff
@@ -632,6 +633,13 @@ static String getMainClassFromJar(String jarname) {
+            if (enableNativeAccess != null) {
+                if (!enableNativeAccess.equals("ALL-UNNAMED")) {
+                    throw new IllegalArgumentException("Only ALL-UNNAMED allowed as value for " + ENABLE_NATIVE_ACCESS);
+                }
```
I don't think throwing IAE is right here. It should call abort with a key for the error message. The value of enableNativeAccess can be used as the parameter for the message.

AlanBateman commented on test/hotspot/jtreg/compiler/rangechecks/TestRangeCheckHoistingScaledIV.java:
```diff
@@ -28,8 +28,8 @@
- * @compile --enable-preview -source ${jdk.version} TestRangeCheckHoistingScaledIV.java
- * @run main/othervm --enable-preview compiler.rangechecks.TestRangeCheckHoistingScaledIV
+ * @compile -source ${jdk.version} TestRangeCheckHoistingScaledIV.java
+ * @run main/othervm compiler.rangechecks.TestRangeCheckHoistingScaledIV
```
Not important but I assume the @compile line can be removed from a number of tests as it's no longer needed. It was needed for tests that didn't use @enablePreview.

JornVernee replied:
Ok, I'll go over all the tests and see if there are `@compile` tags that can be removed

AlanBateman commented on src/java.base/share/classes/java/lang/foreign/MemorySegment.java:
```diff
@@ -1068,52 +1051,123 @@ default long mismatch(MemorySegment other) {
+     *                                       {@linkplain Scope#isAlive() alive}.
+     * @throws WrongThreadException          if this method is called from a thread {@code T},
+     *                                       such that {@code isAccessibleBy(T) == false}.
+     * @throws UnsupportedOperationException if {@code charset} is not a {@linkplain StandardCharsets standard charset}.
```
The caller can fix/avoid the exception by providing another value for the argument so I think IAE is the unchecked exception for this case rather than UOE.

JornVernee replied:
I agree. I'll make the change for the following `CharSet` accepting methods: `MemorySegment::getString(long,Charset)`, `MemorySegment::setString(long,String,Charset)`, and `SegmentAllocator::allocateFrom(String,Charset)`. (Which should be all of them).

ExE-Boss commented:
Maybe include <https://github.com/openjdk/panama-foreign/pull/864>?

JornVernee commented:
> Maybe include [openjdk/panama-foreign#864](https://github.com/openjdk/panama-foreign/pull/864)?

The leftover implementation-only changes will be ported over separately.

JornVernee commented:
I've addressed one more test failure in [b4a7b7a](https://github.com/openjdk/jdk/pull/15103/commits/b4a7b7abe75aee72c8e56c12c0ddbeb597e15f1b) that was found after merging this patch with the most recent master branch.

The test in questions checks for linter warnings for restricted methods. It depends on the FFM API for that, which is still preview in the master branch, but no longer preview as part of this patch. Hence the warnings the compiler outputs is different, and the test fails. I've fixed to test. The preview flags are no longer needed.

JornVernee commented:
/integrate
